### PR TITLE
Bugfix: Providers with batched updates might not report the correct number of changes

### PR DESCRIPTION
**Pull request by tlimoncelli**
* The "number of corrections" was listed incorrectly for many providers.
* Sometimes the count was the # of API calls needed to correct the changes, not the number of corrections.
* For NAMESERVER() corrections, sometimes the count included 1 for a warning

### CODE REVIEW

**Comment by tlimoncelli** [comment]
CC @tresni @riyadhalnur @hnrgrgr @vojtad @das7pad @kallsyms @fabienmazieres @Veratil 

Hey folks!  I'd like you to do some manual testing for me.

As you know, "dnscontrol preview" reports the number of corrections.  However sometimes that count is off. In particular, if the provider permits "batched updates" there's a chance that the number of corrections reported is actually the number of batches.

Is that true?  If so, this mega-PR will fix that.

Could you check your provider to see if that's true and, if so, does this PR fix it?

You'll have to manually create a dnsconfig.js file, since the integration tests don't test this kind of thing.  (though maybe it should?)

I'd appreciate the help.  If you could do this in the next week (by Sept 11) that'd be awesome.

Thanks!
Tom

**Comment by hnrgrgr** [comment]
Looks good with AXFRDDNS!

Old output:
```
> ../dnscontrol/dnscontrol preview  
******************** Domain: example.com
1 correction (axfrddns)
#1: DDNS UPDATES to 'example.com' (primary master: '172.17.0.2:53'). Changes:
± MODIFY ns2.example.com A (172.17.0.2 ttl=300) -> (172.17.0.3 ttl=300)
- DELETE ns3.example.com A 172.17.0.2 ttl=300
+ CREATE example.com NS ns2.example.com. ttl=300
Done. 1 corrections.
```

New output:
```
> ../dnscontrol/dnscontrol ppreview 
CONCURRENTLY gathering 0 zone(s)
SERIALLY gathering 1 zone(s)
Serially Gathering: "example.com"
******************** Domain: example.com
3 corrections (axfrddns)
#1: DDNS UPDATES to 'example.com' (primary master: '172.17.0.2:53'). Changes:
± MODIFY ns2.example.com A (172.17.0.2 ttl=300) -> (172.17.0.3 ttl=300)
- DELETE ns3.example.com A 172.17.0.2 ttl=300
+ CREATE example.com NS ns2.example.com. ttl=300
Done. 3 corrections.

```

**Comment by fabienmazieres** [comment]
Does not looks like Oracle is behaving any differently:

```
1 correction (oracle)
#1: + CREATE A test3.example.co.uk 2.3.4.5 ttl=300
+ CREATE A test4g.example.co.uk 1.1.1.1 ttl=300- DELETE A test2.example.co.uk 2.3.4.5 ttl=300
```

**Comment by tlimoncelli** [comment]
@fabienmazieres I'm surprised there's no change. Is there a chance you're on the old branch or your PATH has the old and new binary?

I've made a special version that fixes a formatting bug.  Please try this:

```
git pull
git checkout tlim_countchanges_oracle
go install
cd to_where_your_dnsconfig_is
dnscontrol preview
```

If that works better, I'll merge it into this PR.

**Comment by tlimoncelli** [comment]
> Looks good with AXFRDDNS!
> 
> Old output:
> 
> ```
> > ../dnscontrol/dnscontrol preview  
> ******************** Domain: example.com
> 1 correction (axfrddns)
> #1: DDNS UPDATES to 'example.com' (primary master: '172.17.0.2:53'). Changes:
> ± MODIFY ns2.example.com A (172.17.0.2 ttl=300) -> (172.17.0.3 ttl=300)
> - DELETE ns3.example.com A 172.17.0.2 ttl=300
> + CREATE example.com NS ns2.example.com. ttl=300
> Done. 1 corrections.
> ```
> 
> New output:
> 
> ```
> > ../dnscontrol/dnscontrol ppreview 
> CONCURRENTLY gathering 0 zone(s)
> SERIALLY gathering 1 zone(s)
> Serially Gathering: "example.com"
> ******************** Domain: example.com
> 3 corrections (axfrddns)
> #1: DDNS UPDATES to 'example.com' (primary master: '172.17.0.2:53'). Changes:
> ± MODIFY ns2.example.com A (172.17.0.2 ttl=300) -> (172.17.0.3 ttl=300)
> - DELETE ns3.example.com A 172.17.0.2 ttl=300
> + CREATE example.com NS ns2.example.com. ttl=300
> Done. 3 corrections.
> ```

Fantastic!  Thanks for the quick reply!

Tom

**Comment by das7pad** [comment]
Hi Tom,

Indeed, the number of corrections was not correct for the HETZNER provider which batches changes. Something is not quite right with the reporting of the new count though, it is still reporting a single change for the batch, despite returning a different `actualChangeCount`.

Greetings,
Jakob

<details>

Before 9178a837de6cdfd4ae658169144f8130bbcb34a7:
```
******************** Domain: testing12.dev
4 corrections (HETZNER)
#1: - DELETE TXT testing12.dev "hello0" ttl=300
#2: - DELETE TXT testing12.dev "hello1" ttl=300
#3: Batch creation of records:
        + CREATE A testing12.dev 1.2.3.42 ttl=300
        + CREATE A testing12.dev 1.2.3.43 ttl=300
#4: Batch modification of records:
        ± MODIFY A testing12.dev: (1.2.3.50 ttl=300) -> (1.2.3.40 ttl=300)
        ± MODIFY A testing12.dev: (1.2.3.51 ttl=300) -> (1.2.3.41 ttl=300)
Done. 4 corrections.
``` 
Print `actualChangeCount`:
```diff
diff --git a/providers/hetzner/hetznerProvider.go b/providers/hetzner/hetznerProvider.go
index 8f911919..1c44bfba 100644
--- a/providers/hetzner/hetznerProvider.go
+++ b/providers/hetzner/hetznerProvider.go
@@ -136,6 +136,7 @@ func (api *hetznerProvider) GetZoneRecordsCorrections(dc *models.DomainConfig, e
                corrections = append(corrections, corr)
        }
 
+       fmt.Printf("actualChangeCount=%d\n", actualChangeCount)
        return corrections, actualChangeCount, nil
 }
 
```
After dbb353adb36c675f17989da0249f8a22c364f6f6 with the above patch:
```
******************** Domain: testing12.dev
actualChangeCount=6
4 corrections (HETZNER)
#1: - DELETE TXT testing12.dev "hello0" ttl=300
#2: - DELETE TXT testing12.dev "hello1" ttl=300
#3: Batch creation of records:
        + CREATE A testing12.dev 1.2.3.42 ttl=300
        + CREATE A testing12.dev 1.2.3.43 ttl=300
#4: Batch modification of records:
        ± MODIFY A testing12.dev: (1.2.3.50 ttl=300) -> (1.2.3.40 ttl=300)
        ± MODIFY A testing12.dev: (1.2.3.51 ttl=300) -> (1.2.3.41 ttl=300)
Done. 4 corrections.
``` 
</details>

**Comment by Veratil** [comment]
Output on the branch, ensured that the PowerDNS changes were in place as well. Still shows only 1 correction with 25 changes.
```
$ dnscontrol preview
******************** Domain: test2.internal
1 correction (powerdns)
#1: ± BATCHED CHANGE/CREATEs for test2.internal
+ CREATE a.test2.internal A 127.0.0.1 ttl=300
+ CREATE b.test2.internal A 127.0.0.2 ttl=300
+ CREATE c.test2.internal A 127.0.0.3 ttl=300
+ CREATE d.test2.internal A 127.0.0.4 ttl=300
+ CREATE e.test2.internal A 127.0.0.5 ttl=300
+ CREATE f.test2.internal A 127.0.0.6 ttl=300
+ CREATE g.test2.internal A 127.0.0.7 ttl=300
+ CREATE h.test2.internal A 127.0.0.8 ttl=300
+ CREATE i.test2.internal A 127.0.0.9 ttl=300
+ CREATE j.test2.internal A 127.0.0.10 ttl=300
+ CREATE k.test2.internal A 127.0.0.11 ttl=300
+ CREATE l.test2.internal A 127.0.0.12 ttl=300
+ CREATE m.test2.internal A 127.0.0.13 ttl=300
+ CREATE n.test2.internal A 127.0.0.14 ttl=300
+ CREATE o.test2.internal A 127.0.0.15 ttl=300
+ CREATE p.test2.internal A 127.0.0.16 ttl=300
+ CREATE q.test2.internal A 127.0.0.17 ttl=300
+ CREATE r.test2.internal A 127.0.0.18 ttl=300
+ CREATE s.test2.internal A 127.0.0.19 ttl=300
+ CREATE t.test2.internal A 127.0.0.20 ttl=300
+ CREATE u.test2.internal A 127.0.0.21 ttl=300
+ CREATE v.test2.internal A 127.0.0.22 ttl=300
+ CREATE w.test2.internal A 127.0.0.23 ttl=300
+ CREATE x.test2.internal A 127.0.0.24 ttl=300
+ CREATE y.test2.internal A 127.0.0.25 ttl=300
WARNING: No nameservers declared; skipping registrar. Add {no_ns:'true'} to force.
Done. 1 corrections.
```

Tracing this back it's because the change count was discarded. Starting at the end of `CorrectZoneRecords`:
```
(dlv) n
> github.com/StackExchange/dnscontrol/v4/pkg/zonerecs.CorrectZoneRecords() /home/veratil/github/dnscontrol/pkg/zonerecs/zonerecords.go:38 (PC: 0x9af9c1)
Warning: debugging optimized function
    33:         dc.Punycode()
    34:         // FIXME(tlim) It is a waste to PunyCode every iteration.
    35:         // This should be moved to where the JavaScript is processed.
    36:
    37:         everything, actualChangeCount, err := driver.GetZoneRecordsCorrections(dc, existingRecords)
=>  38:         reports, corrections := splitReportsAndCorrections(everything)
    39:         return reports, corrections, actualChangeCount, err
    40: }
    41:
    42: func splitReportsAndCorrections(everything []*models.Correction) (reports, corrections []*models.Correction) {
    43:         for i := range everything {
(dlv) p actualChangeCount
25
```
But then in `commands/previewPush.go` the value is discarded:
```
> github.com/StackExchange/dnscontrol/v4/commands.run.func1() /home/veratil/github/dnscontrol/commands/previewPush.go:250 (PC: 0xa9b6f8)
Warning: debugging optimized function
Values returned:
        ~r0: (unreadable empty OP stack)
        ~r1: (unreadable empty OP stack)
        ~r2: (unreadable empty OP stack)
        ~r3: (unreadable empty OP stack)

   245:                                 out.StartDNSProvider(provider.Name, !shouldrun)
   246:                                 if !shouldrun {
   247:                                         continue
   248:                                 }
   249:
=> 250:                                 reports, corrections, _, err := zonerecs.CorrectZoneRecords(provider.Driver, domain)
   251:                                 out.EndProvider(provider.Name, len(corrections), err)
   252:                                 if err != nil {
   253:                                         anyErrors = true
   254:                                         return
   255:                                 }
```

**Comment by tlimoncelli** [comment]
Bingo!  I think you found it!

I had fixed `ppreview` but not `preview`.  I got interrupted and when I returned  I forgot to go back and finish `preview`.

Thank you for finding my mistake!

Could you try again with `ppreview` to see if the count works there?

**Comment by Veratil** [comment]
```
$ dnscontrol ppreview
CONCURRENTLY gathering 0 zone(s)
SERIALLY gathering 1 zone(s)
Serially Gathering: "test2.internal"
******************** Domain: test2.internal
25 corrections (powerdns)
#1: ± BATCHED CHANGE/CREATEs for test2.internal
+ CREATE a.test2.internal A 127.0.0.1 ttl=300
+ CREATE b.test2.internal A 127.0.0.2 ttl=300
+ CREATE c.test2.internal A 127.0.0.3 ttl=300
+ CREATE d.test2.internal A 127.0.0.4 ttl=300
+ CREATE e.test2.internal A 127.0.0.5 ttl=300
+ CREATE f.test2.internal A 127.0.0.6 ttl=300
+ CREATE g.test2.internal A 127.0.0.7 ttl=300
+ CREATE h.test2.internal A 127.0.0.8 ttl=300
+ CREATE i.test2.internal A 127.0.0.9 ttl=300
+ CREATE j.test2.internal A 127.0.0.10 ttl=300
+ CREATE k.test2.internal A 127.0.0.11 ttl=300
+ CREATE l.test2.internal A 127.0.0.12 ttl=300
+ CREATE m.test2.internal A 127.0.0.13 ttl=300
+ CREATE n.test2.internal A 127.0.0.14 ttl=300
+ CREATE o.test2.internal A 127.0.0.15 ttl=300
+ CREATE p.test2.internal A 127.0.0.16 ttl=300
+ CREATE q.test2.internal A 127.0.0.17 ttl=300
+ CREATE r.test2.internal A 127.0.0.18 ttl=300
+ CREATE s.test2.internal A 127.0.0.19 ttl=300
+ CREATE t.test2.internal A 127.0.0.20 ttl=300
+ CREATE u.test2.internal A 127.0.0.21 ttl=300
+ CREATE v.test2.internal A 127.0.0.22 ttl=300
+ CREATE w.test2.internal A 127.0.0.23 ttl=300
+ CREATE x.test2.internal A 127.0.0.24 ttl=300
+ CREATE y.test2.internal A 127.0.0.25 ttl=300
1 correction (none)
INFO#1: No nameservers declared for domain "test2.internal"; skipping registrar. Add {no_ns:'true'} to force
Done. 26 corrections.
```
Two things to notice:
1. `1 correction (none)` which I don't know what's going on there
2. Off by one error with `26 corrections`, I assume this is from the empty `1 correction (none)`

Looks like this happens here:
```
> github.com/StackExchange/dnscontrol/v4/commands.prun() /home/veratil/github/dnscontrol/commands/ppreviewPush.go:258 (PC: 0xa95179)
Warning: debugging optimized function
   253:                 out.StartRegistrar(zone.RegistrarName, !skip)
   254:                 if skip {
   255:                         corrections := zone.GetCorrections(zone.RegistrarInstance.Name)
   256:                         numActions := zone.GetChangeCount(zone.RegistrarInstance.Name)
   257:                         out.EndProvider2(zone.RegistrarName, numActions)
=> 258:                         totalCorrections += numActions
   259:                         reportItems = append(reportItems, genReportItem(zone.Name, corrections, zone.RegistrarName))
   260:                         anyErrors = cmp.Or(anyErrors, pprintOrRunCorrections(zone.Name, zone.RegistrarInstance.Name, corrections, out, push, interactive, notifier, report))
   261:                 }
   262:
   263:         }
(dlv) p numActions
1
(dlv) p totalCorrections
25
(dlv) n
> github.com/StackExchange/dnscontrol/v4/commands.prun() /home/veratil/github/dnscontrol/commands/ppreviewPush.go:259 (PC: 0xa951b5)
Warning: debugging optimized function
   254:                 if skip {
   255:                         corrections := zone.GetCorrections(zone.RegistrarInstance.Name)
   256:                         numActions := zone.GetChangeCount(zone.RegistrarInstance.Name)
   257:                         out.EndProvider2(zone.RegistrarName, numActions)
   258:                         totalCorrections += numActions
=> 259:                         reportItems = append(reportItems, genReportItem(zone.Name, corrections, zone.RegistrarName))
   260:                         anyErrors = cmp.Or(anyErrors, pprintOrRunCorrections(zone.Name, zone.RegistrarInstance.Name, corrections, out, push, interactive, notifier, report))
   261:                 }
   262:
   263:         }
   264:
(dlv) p totalCorrections
26
```

**Comment by tlimoncelli** [comment]
> 1. 1 correction (none)

The `NONE` provider is a provider that does nothing.  Not sure why it counts a correction, but I can investigate.

**Comment by riyadhalnur** [comment]
Tested with gcloud and this PR fixes the issue with preview for batched updates

With `preview`
```shell
> go run main.go preview
******************** Domain: dnscontrol-gcloud.com
1 correction (gcloud)
#1: + CREATE test1.dnscontrol-gcloud.com A 1.2.3.4 ttl=300
+ CREATE test2.dnscontrol-gcloud.com A 5.6.7.8 ttl=300
Done. 1 corrections.
```

With `ppreview`
```shell
> go run main.go ppreview
CONCURRENTLY gathering 1 zone(s)
SERIALLY gathering 0 zone(s)
Waiting for concurrent gathering(s) to complete...DONE
******************** Domain: dnscontrol-gcloud.com
2 corrections (gcloud)
#1: + CREATE test1.dnscontrol-gcloud.com A 1.2.3.4 ttl=300
+ CREATE test2.dnscontrol-gcloud.com A 5.6.7.8 ttl=300
Done. 2 corrections.
````

**Comment by tlimoncelli** [comment]
Hey folks!  I've fixed `preview` and now the counts work the same as `ppreview`.

@Veratil: I haven't been able to repro the  `1 correction (none)` statement. Can you help me with that?  Can you send me (a redacted) dnsconfig.js?

**Comment by Veratil** [comment]
> @Veratil: I haven't been able to repro the `1 correction (none)` statement. Can you help me with that? Can you send me (a redacted) dnsconfig.js?

Use `go-powerdns`s docker compose to start up a powerdns container, api key is in the compose file for testing. Test dnsconfig used:
```
var REG_NONE = NewRegistrar("none");
var PROV_PDNS = NewDnsProvider("powerdns");

var RRSETS = [];

var alphabet = ["a","b","c","d","e","f","g","h","i","j","k","l","m","n","o","p","q","r","s","t","u","v","w","x","y","z"];
var offset = 0;
while (alphabet[offset] != 'z') {
        RRSETS.push(A(alphabet[offset], "127.0.0." + (offset + 1).toString()));
        offset += 1;
};

D("test2.internal", REG_NONE, DnsProvider(PROV_PDNS),
        RRSETS,
END);
```

**Comment by tlimoncelli** [comment]
@Veratil Thanks!  I can repro it now.  Fix should be coming soon.

**Comment by tlimoncelli** [comment]
@Veratil Fixed!  

Old and broken:

```
tlimworkm3:~/tmp/cloudnsTest$ dnscontrol preview && dnscontrol ppreview
******************** Domain: test2.internal
WARNING: No nameservers declared; skipping registrar. Add {no_ns:'true'} to force.
Done. 0 corrections.
CONCURRENTLY gathering 1 zone(s)
SERIALLY gathering 0 zone(s)
Waiting for concurrent gathering(s) to complete...DONE
******************** Domain: test2.internal
1 correction (none)
INFO#1: No nameservers declared for domain "test2.internal"; skipping registrar. Add {no_ns:'true'} to force
Done. 1 corrections.
```

New and working:

```
tlimworkm3:~/tmp/cloudnsTest$ dnscontrol preview && dnscontrol ppreview
******************** Domain: test2.internal
WARNING: No nameservers declared; skipping registrar. Add {no_ns:'true'} to force.
Done. 0 corrections.
CONCURRENTLY gathering 1 zone(s)
SERIALLY gathering 0 zone(s)
Waiting for concurrent gathering(s) to complete...DONE
******************** Domain: test2.internal
INFO#1: No nameservers declared for domain "test2.internal"; skipping registrar. Add {no_ns:'true'} to force
Done. 0 corrections.
```

If I don't hear any complaints, I'll probably merge this on Sunday or Monday.

File: commands/ppreviewPush.go [hunk 1]
```diff
@@ -543,24 +543,24 @@ func generateZoneCorrections(zone *models.DomainConfig, provider *models.DNSProv
 	return zoneCorrections, reports, actualChangeCount
 }
 
-func generateDelegationCorrections(zone *models.DomainConfig, providers []*models.DNSProviderInstance, _ *models.RegistrarInstance) []*models.Correction {
+func generateDelegationCorrections(zone *models.DomainConfig, providers []*models.DNSProviderInstance, _ *models.RegistrarInstance) ([]*models.Correction, int) {
 	//fmt.Printf("DEBUG: generateDelegationCorrections start zone=%q nsList = %v\n", zone.Name, zone.Nameservers)
 	nsList, err := nameservers.DetermineNameserversForProviders(zone, providers, true)
 	if err != nil {
-		return msg(fmt.Sprintf("DtermineNS: zone %q; Error: %s", zone.Name, err))
+		return msg(fmt.Sprintf("DtermineNS: zone %q; Error: %s", zone.Name, err)), 0
```
Comment (Veratil) [inline commands/ppreviewPush.go:550]: Out of curiosity, is `DtermineNS` with the missing `e` right?

Comment (tlimoncelli) [reply commands/ppreviewPush.go]: Good catch!  I've fixed it.

File: models/domain.go [hunk 2]
```diff
@@ -176,3 +177,23 @@ func (dc *DomainConfig) GetCorrections(providerName string) []*Correction {
 	}
 	return nil
 }
+
+// IncrementChangeCount accumulates change cound in a thread-safe way.
```
Comment (riyadhalnur) [inline models/domain.go:181]: Typo `cound` -> `count`

File: models/domain.go [hunk 3]
```diff
@@ -176,3 +177,23 @@ func (dc *DomainConfig) GetCorrections(providerName string) []*Correction {
...
+	defer dc.pendingCorrectionsMutex.Unlock()
+
+	if dc.pendingActualChangeCount == nil {
+		// First time storing anything.
+		dc.pendingActualChangeCount = make(map[string](int))
+	}
+	dc.pendingActualChangeCount[providerName] += delta
+}
+
+// GetChangeCount accumulates change cound in a thread-safe way.
```
Comment (riyadhalnur) [inline models/domain.go:193]: Same typo as above here

**Review by riyadhalnur — APPROVED** [review]

File: models/domain.go (hunk 3 above)
Comment (tlimoncelli) [reply models/domain.go:193]: Thanks!

**Comment by tlimoncelli** [comment]
Thanks for all your help!  I couldn't have done it without all of your input!

**Pull request merged**